Add new post-store callback API

mistydemeo opened the pull request:
This adds support for post-store callbacks, via a new API that would be called by Archivematica. The current usecase is to support reporting back to Islandora for each file that's stored.

The storage service currently doesn't maintain a persistent memory of files; this adds a new minimal File model that can be used to track files on transfer start. Since we don't currently assign or track UUIDs to individual files in the storage service, this instead calculates checksums on arrival and uses that to identify the file at a later point in time. This is kind of a hack.

(In the long term, I think it'd make sense to do some of the file cleanup and UUID assignment in the storage service, and send the files over to Archivematica with UUIDs assigned and the transfer METS already built; then we can just inspect the AIP METS to locate files. That would be best done when we have a) a background task scheduler, b) a standalone METS writer, though.)

The settings for the callback are global; may make more sense to make them controlled per-pipeline.

Another few supporting improvements:
- The SWORD helpers checksum function has been replaced with a checksum function in common.utils that uses hashlib, and supports any checksum algorithm that hashlib supports. File's checksum column was sized large enough to fit a whirlpool or sha512 hash<s>,though that's bigger than we'd actually use for the time being</s>. UPDATE: Storing sha512 now.
- Package gained a new instance method, `Package.each_file`, that iterates over each file in a package as a convenience wrapper around `os.walk`. It also supports extracting compressed packages if necessary.

This isn't tested against an actual Islandora instance yet.

Commit:
SWORD: record each file to the database

In the absense of persistent UUIDs originating in the storage service,
storing the checksum and source ID allows the file to be recognized
when it comes back around again.

Commit:
Add post-store AIP callbacks

This adds a new REST endpoint that allows a post-store callback to be
executed for every file in a given AIP.

This will only trigger for files which exist in the File table, which
in practice means only files that were transferred from Islandora.

refs #6788

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -524,6 +526,63 @@ def pointer_file_request(self, request, **kwargs):
+        fail = 0
+
+        for f in package.each_file(extract=True):
+            cksum = utils.get_file_checksum(f, 'sha256')
```
Can probably fetch this from the bag's manifest if it exists, and only fall back to calculating it if not.

mistydemeo replied:
We're storing sha512; I'll switch to calculating those on creation and request file info from the bag.

mistydemeo commented:
Since I'm no longer using it, may delete `Package.each_file`.

Hwesta commented on storage_service/common/utils.py:
```diff
@@ -102,6 +103,21 @@ def download_file_stream(filepath, temp_dir=None):
 
 ############ OTHER ############
 
+def get_file_checksum(filepath, algorithm):
```
This was also added in d03e4e59dad82fd810b058ce7cd91efde2bc2c19, which returns the checksum object instead of the hexdigest.  One will get replaced when this gets rebased.

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -524,6 +528,79 @@ def pointer_file_request(self, request, **kwargs):
 
+    def aip_store_callback_request(self, request, **kwargs):
+        package = get_object_or_None(Package, uuid=kwargs['uuid'])
+        if package and package.package_type != Package.AIP:
```
What if package is None?

mistydemeo replied:
Updated to use the request endpoint instead. (Didn't exist where this was originally branched from.)

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -524,6 +528,79 @@ def pointer_file_request(self, request, **kwargs):
 
         return response
 
+    def aip_store_callback_request(self, request, **kwargs):
```
Should be updated to use @_custom_endpoint decorator, once rebased (which will also handle the case of the Package being None)

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -524,6 +528,79 @@ def pointer_file_request(self, request, **kwargs):
+            uri = base_callback_uri.replace('<source_id>', source_id)
+            try:
+                requests.post(uri)
+            except:
```
Catch a more specific exception (even just Exception).  Maybe log,  `logging.info(..., exc_info=True)`?

Commit:
sword/helpers: replace custom md5 function

mistydemeo commented:
Rebased and reworked on the Fedora class.

Now introduces a new `Callback` class, which allows for generic callbacks to be associated with given types of events. The only such event right now is "post-store", but this could be used other places. For instance, AIP reingest could use this instead of using its custom callback.

This also accommodates multiple callbacks for a given event, so it's possible to POST reports to multiple services.

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,61 @@ def check_fixity_request(self, request, bundle, **kwargs):
+        fail = 0
+
+        if package.is_compressed():
+            package_dir, tmpdir = package.extract_file()
```
Possible optimization: see if I can extract just the bag service files and not the bag contents for a quicker way to access the bag properties from the manifest.

Hwesta replied:
extract_file takes a parameter for the relative path within the package - you could use that to just get the service file(s)

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,61 @@ def check_fixity_request(self, request, bundle, **kwargs):
+        if fail > 0:
+            response = {
+                "message": "Failed to POST {} responses to callback URI".format(fail),
+                "failure_count": fail,
```
Now that there's multiple URLs, failures should probably be more specific, e.g. a list of:

``` python
{
    "file": "filename",
    "callback_uri": "uri"
}
```

mistydemeo commented:
Updated with forms to create new callbacks, and a migration for callbacks.

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,61 @@ def check_fixity_request(self, request, bundle, **kwargs):
+            except KeyError:
+                # These files do not typically have an sha512 hash, so it's
+                # fine for these to be missing that key; every other file should.
+                safe_files = ('bag-info.txt', 'manifest-sha512.txt', 'bagit.txt')
```
Define this outside the loop?

mistydemeo replied:
Good point.

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,61 @@ def check_fixity_request(self, request, bundle, **kwargs):
+                safe_files = ('bag-info.txt', 'manifest-sha512.txt', 'bagit.txt')
+                if f not in safe_files:
+                    logging.warning("Post-store callback: sha512 missing for file %s", f)
+                    continue
```
What if `f in safe_files`?  Won't `cksum` be undefined below?

mistydemeo replied:
This isn't a subloop - it'll continue to processing the next file, and won't try to do anything with this one.

Hwesta replied:
I'm thinking this case:

``` python
for f, checksums in bag.entries.iteritems():
    try:
        cksum = checksums['sha512']  # Raises exception
    except KeyError:
        safe_files = ('bag-info.txt', 'manifest-sha512.txt', 'bagit.txt')

        if f not in safe_files:  # f is in safe_files, does not trigger this
            logging.warning("Post-store callback: sha512 missing for file %s", f)
            continue
    try:
        # Continues to here - cksum not defined because exception raised,
        # continue not hit because f in safe_files
        source_id = File.objects.get(checksum=cksum).source_id
```

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,61 @@ def check_fixity_request(self, request, bundle, **kwargs):
+                    continue
+            try:
+                source_id = File.objects.get(checksum=cksum).source_id
+            except File.DoesNotExist:
```
Catch `File.MultipleObjectsReturned` as well?

mistydemeo replied:
Good point!

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,61 @@ def check_fixity_request(self, request, bundle, **kwargs):
+        bag = bagit.Bag(package_dir)
+        for f, checksums in bag.entries.iteritems():
+            try:
+                cksum = checksums['sha512']
```
Nice trick, using the existing bagit checksums instead of recalculating.

mistydemeo commented on requirements/base.txt:
```diff
@@ -7,6 +7,7 @@ South==0.8.4
 django-tastypie==0.9.15
 django-extensions==1.1.1
 lxml==3.2.3
+requests==2.2.1
```
Should we look at upgrading to requests 2.3.1 across the board? I used 2.2.1 because another branch is using the same version.

Hwesta replied:
Sure - do we want to just specify `requests>=2.2.1?`?

mistydemeo replied:
:thumbsup:

Hwesta commented:
Minor: The commit order of adding models (File, Callback) should come before they're used.

Do we want to merge the Callback and File migrations?  They'll be coming in together, and are conceptually one feature.  To keep individual commits working, the later migration could use --update.

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -332,3 +332,47 @@ def space_delete(request, uuid):
+    except Callback.DoesNotExist:
+        messages.warning(request, "Callback {} does not exist.".format(location_uuid))
+        return redirect('callback_list')
+    return render(request, 'locations/callback_detail.html', locals())
```
I don't see the templates committed...

mistydemeo commented:
Committed callback templates - looks like those got missed!

mistydemeo commented:
> Do we want to merge the Callback and File migrations?

Makes sense, I think. Planning to merge those together when this branch gets merged; right now they're separate just because they represent the order in which I wrote features in the branch.

mistydemeo commented:
> Minor: The commit order of adding models (File, Callback) should come before they're used.

Github is showing the incorrect order. The order is:
- sword/helpers md5
- callback model
- file model
- commits those use those models

Hwesta commented:
> > Minor: The commit order of adding models (File, Callback) should come before they're used.
> 
> Github is showing the incorrect order. The order is:
>    sword/helpers md5
>    callback model
>    file model
>    commits those use those models
> Okay

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -332,3 +332,47 @@ def space_delete(request, uuid):
+
+def callback_switch_enabled(request, uuid):
+    callback = get_object_or_404(Callback, uuid=uuid)
+    callback.enabled = not callback.enabled
```
Are [F()](https://docs.djangoproject.com/en/dev/ref/models/queries/#f-expressions) expressions overkill here? Would prevent race conditions.  (Discovered them recently, which is why they don't show up elsewhere...)

mistydemeo replied:
The logic here is copied from other places, so if we change them to use `F()` expressions let's do it out of this PR.

Hwesta commented on storage_service/locations/views.py:
```diff
@@ -332,3 +332,47 @@ def space_delete(request, uuid):
+            return redirect('callback_detail', callback.uuid)
+    else:
+        form = forms.CallbackForm(instance=callback)
+        return render(request, 'locations/callback_form.html', locals())
```
The previous 9 lines can be simplified to

``` python
form = forms.CallbackForm(request.POST or None, instance=callback)
if form.is_valid():
    callback = form.save()
    messages.success(request, "Callback saved.")
    return redirect('callback_detail', callback.uuid)
return render(request, 'locations/callback_form.html', locals())
```

which I think is a really neat trick.

Hwesta commented:
Currently we only checksum and add files to the Files table for FEDORA stuff.  That means if someone creates an AIP not using the SWORD API, and sets up a callback, it will silently fail trying to match files, correct? Is this a problem? Should we try to support more general usecases, since the callback structure is more generic?  Post a warning that this is the case?

Commit:
Space: update_package_status must always return a tuple

Commit:
SWORD helpers: name Fedora METS after object ID

Commit:
SWORD API: store all METS files

Previously, only the first METS file for a given set of files would be
stored. This stores every METS file, so that multi-file transfers will
contain the full set of metadata.

Commit:
SWORD: download MODS records with OBJ items

Hwesta commented on storage_service/locations/api/resources.py:
```diff
@@ -633,6 +635,65 @@ def check_fixity_request(self, request, bundle, **kwargs):
+    def aip_store_callback_request(self, request, bundle, **kwargs):
+        package = bundle.obj
+
+        callbacks = Callback.objects.filter(event="post_store")
```
Might make sense to have the type of event included in the callback, so we can reuse this endpoint. Ie, instead of calling `send_store_callback` and having the endpoint know,  do `send_callback/post_store/` or `send_callback` with the body having {"type": "post_store"}, so it's more generic.

mistydemeo replied:
Changed the URL to `send_callback/post_store/`, but the logic is too tightly coupled to the post-store AIP usecase to genericize the function logic.

Hwesta commented on storage_service/locations/migrations/0006_auto__add_callback.py:
```diff
@@ -0,0 +1,220 @@
+# -*- coding: utf-8 -*-
```
Merge the two migrations into one file and give it a more meaningful name?

mistydemeo replied:
Merged into `0005_v0_5_0.py`.

Commit:
Add File, Callback models

Commit:
Add Callback forms, views

This allows for generic REST API callbacks to be associated with given
types of events. A callback is given a URL pattern, an HTTP method to
use when contacting that URL, and an expected HTTP status.

The only such event right now is "post-store", but this could be used
other places. For instance, AIP reingest could use this instead of
using its custom callback.

refs #6788

Commit:
Package: extract base directory logic

mistydemeo commented:
Updated the post-store callback to not extract the entire AIP, and added a new function in 118d9bafd56fda66b592040b96559a43f53173d5 to get the directory basename.

Going to test this to make sure it actually _works_, given the number of changes and rebasing.
